src/repr/src/row.rs

```diff
@@ -1193,6 +1328,161 @@ impl RowPacker<'_> {
+            },
+        )
+    }
+
+    /// Pushes a `DatumRange` built from the specified arguments, and
+    /// permitting the user to push values into the row using `f`.
+    ///
+    /// - If an error occurs, the packer's state will be unchanged.
+    /// - Prefer using [`Self::push_range`], which prevents most potential
+    ///   panics.
+    /// - Infinite boundaries should be expressed using [`RangeBoundDesc`]'s
+    ///   `inf` field.
+    ///
+    ///   Note that a future version of this API might panic on pushing
+    ///   `Datum::Null` values.
+    /// - `f` only supports operations which push datums (`push_datum`, etc.).
+    ///   Any other usage results in undefined, unsafe behavior.
+    /// - To construct an empty range, use [`Self::push_empty_range`]. Note
+    ///   that it is possible to generate semantically empty ranges with this
+    ///   function, but they will _not_ be output as empty ranges.
+    ///
+    /// # Panics
+    /// - If any `Datum::Null` were pushed using `f`.
+    /// - If the number of `Datum`s pushed using `f` is incommensurate with the
+    ///   provided `RangeBoundDesc` or the structure of ranges in general.
+    /// - If `Datum`s of different types were pushing using `f`.
+    pub fn push_range_with<F, R>(
+        &mut self,
+        RangeBoundDesc {
+            inf: lower_inf,
+            inc: lower_inc,
+        }: RangeBoundDesc,
+        RangeBoundDesc {
+            inf: upper_inf,
+            inc: upper_inc,
+        }: RangeBoundDesc,
+        f: F,
+    ) -> Result<R, RangeError>
+    where
+        F: FnOnce(&mut RowPacker) -> R,
+    {
+        let start = self.row.data.len();
+        self.row.data.push(Tag::Range.into());
+
+        let mut flags: u8 = 0;
+        let mut expected_datums = 0;
+
+        for (val, set, expected_datum_diff) in [
+            (lower_inf, range::RANGE_LB_INF, 1),
+            (lower_inc, range::RANGE_LB_INC, 0),
+            (upper_inf, range::RANGE_UB_INF, 1),
+            (upper_inc, range::RANGE_UB_INC, 0),
+        ] {
+            if val {
+                flags |= set;
+            } else {
+                expected_datums += expected_datum_diff;
+            }
+        }
+
+        self.row.data.push(flags);
+        let mut datum_check = self.row.data.len();
+
+        let out = f(self);
+
+        // Validate that what was written maintains the correct invariants.
+        let mut actual_datums = 0;
+        let mut seen = None;
+        while datum_check < self.row.data.len() {
+            let d = unsafe { read_datum(&self.row.data, &mut datum_check) };
+            match seen {
+                None => seen = Some(d),
+                Some(seen) => {
+                    let seen_kind = DatumKind::from(seen);
+                    let d_kind = DatumKind::from(d);
+                    if seen_kind != d_kind {
+                        panic!("range contains inconsistent data; expected {seen_kind:?} but got {d_kind:?}");
+                    }
```

Comment:
Take it or leave it, but I find the `assert!` macro to be marginally more descriptively named here, and you can provide a custom assertion string to make the error message more readable.

```suggestion
                    assert!(seen_kind == d_kind, "range contains inconsistent data; expected {seen_kind:?} but got {d_kind:?}");
```

Reply:
Ty for catching this; these were all errors instead of panics, but forgot to move to `assert` when I removed the errors.